fronts-client/src/bundles/notificationsBundle.ts

```diff
@@ -0,0 +1,82 @@
+import { Action } from 'types/Action';
+import v4 from 'uuid/v4';
+
+import { State } from 'types/State';
+
+interface BannerNotification {
+  id: string;
+  message: string;
+  duplicates: number;
+}
+
+export interface NotificationState {
+  banners: BannerNotification[];
+}
+
+// Actions
+
+export const NOTIFICATION_ADD_BANNER = 'NOTIFICATION_ADD_BANNER' as const;
+
+export const actionAddNotificationBanner = (message: string) => ({
+  type: NOTIFICATION_ADD_BANNER,
+  payload: { message, id: v4() }
+});
+
+export const NOTIFICATION_REMOVE_BANNER = 'NOTIFICATION_REMOVE_BANNER' as const;
+
+export const actionRemoveNotificationBanner = (id: string) => ({
+  type: NOTIFICATION_REMOVE_BANNER,
+  payload: { id }
+});
+
+export type NotificationActions =
+  | ReturnType<typeof actionAddNotificationBanner>
+  | ReturnType<typeof actionRemoveNotificationBanner>;
+
+// Selectors
+
+export const selectBanners = (state: State) => state.notifications.banners;
+
+// Reducer
+
+export const initialState: NotificationState = { banners: [] };
+
+export const reducer = (
+  state: NotificationState = initialState,
+  action: Action
+): NotificationState => {
+  switch (action.type) {
+    case NOTIFICATION_ADD_BANNER: {
+      const duplicateNotification = state.banners.find(
+        _ => _.message === action.payload.message
+      );
+      if (duplicateNotification) {
+        return {
+          banners: [
+            ...state.banners.filter(_ => _.id !== duplicateNotification.id),
+            {
+              ...duplicateNotification,
+              duplicates: duplicateNotification.duplicates + 1
+            }
+          ]
+        };
+      }
+
+      return {
+        banners: [
+          ...state.banners,
+          {
+            id: action.payload.id,
+            message: action.payload.message,
+            duplicates: 1
```

Comment:
If there is 1 duplicate, does that mean that only a single instance of the notification has been added?

Reply:
Yes, the naming here isn't quite right. Perhaps `instances`?

Reply:
Perhaps stick with the name and default to `0` then. I think that would make sense!